File: Subarray with given sum - GFG/subarray-with-given-sum.java
//{ Driver Code Starts
import java.util.*;
import java.lang.*;
import java.io.*;

class Main{
	static BufferedReader br;
    static PrintWriter ot;
    public static void main(String[] args) throws IOException{
        
        br = new BufferedReader(new InputStreamReader(System.in));
        ot = new PrintWriter(System.out);

        int t = Integer.parseInt(br.readLine());

        while(t-->0){
            
            String s[] = br.readLine().trim().split(" ");
            
            int n = Integer.parseInt(s[0]);
            int k = Integer.parseInt(s[1]);
            int a[] = new int[n];
            s = br.readLine().trim().split(" ");
            for(int i = 0; i < n; i++)
                a[i] = Integer.parseInt(s[i]);
            Solution obj = new Solution();
            ArrayList<Integer> res = obj.subarraySum(a, n, k);
            for(int ii = 0;ii<res.size();ii++)
                ot.print(res.get(ii) + " ");
            ot.println();
        }
        ot.close();
    }

}
// } Driver Code Ends


class Solution
{
    //Function to find a continuous sub-array which adds up to a given number.
    static ArrayList<Integer> subarraySum(int[] arr, int n, int s) 
    {
        // Your code here
        ArrayList<Integer> al = new ArrayList<Integer>();
        boolean found=false;
        int st=0; int end=0;
        int sum=0;
        for(int i=0; i<n; i++)
        {
           sum+=arr[i];
           
           if(sum>=s)
           {
               end=i;
               while(s<sum && st<end)
               {
                   sum-=arr[st++];
                   
               }
           }
           if(sum==s)
           {
               al.add(st+1); al.add(end+1);
               found=true;
               break;
           }
        }
        if(!found)
        al.add(-1);
        
        return al;
        
    }
}

File: Subarray with given sum - GFG/subarray-with-given-sum.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>
using namespace std;


// } Driver Code Ends

class Solution
{
    public:
    //Function to find a continuous sub-array which adds up to a given number.
    vector<int> subarraySum(vector<int>arr, int n, long long s)
    {
         vector<int> result;
    bool found = false;
    int st = 0, end = 0;
    int sum = 0;
    
    for (int i = 0; i < n; i++) {
        sum += arr[i];
        
        if (sum >= s) {
            end = i;
            while (s < sum && st < end) {
                sum -= arr[st++];
            }
        }
        
        if (sum == s) {
            result.push_back(st + 1);
            result.push_back(end + 1);
            found = true;
            break;
        }
    }
    
    if (!found)
        result.push_back(-1);
    
    return result;
    }
};


//{ Driver Code Starts.

int main()
 {
    int t;
    cin>>t;
    while(t--)
    {
        int n;
        long long s;
        cin>>n>>s;
        vector<int>arr(n);
        // int arr[n];
        const int mx = 1e9;
        for(int i=0;i<n;i++)
        {
            cin>>arr[i];
        }
        Solution ob;
        vector<int>res;
        res = ob.subarraySum(arr, n, s);
        
        for(int i = 0;i<res.size();i++)
            cout<<res[i]<<" ";
        cout<<endl;
        
    }
	return 0;
}
// } Driver Code Ends